hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreContext.java

```diff
@@ -0,0 +1,194 @@
+import org.apache.hadoop.hbase.io.HeapSize;
+import org.apache.hadoop.hbase.io.crypto.Encryption;
+import org.apache.hadoop.hbase.io.hfile.CacheConfig;
+import org.apache.hadoop.hbase.util.ClassSize;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * This carries the immutable information and references on some of the meta data about the HStore.
+ * This meta data can be used across the HFileWriter/Readers and other HStore consumers without the
+ * need of passing around the complete store.
+ */
+@InterfaceAudience.Private
+public final class StoreContext implements HeapSize {
+  public static final long FIXED_OVERHEAD = ClassSize.estimateBase(HStore.class, false);
+
+  private final int blockSize;
+  private final Encryption.Context encryptionContext;
+  private final CacheConfig cacheConf;
+  private final HRegionFileSystem regionFileSystem;
+  private final CellComparator comparator;
+  private final BloomType bloomFilterType;
+  private final Supplier<Collection<HStoreFile>> compactedFilesSupplier;
+  private final Supplier<InetSocketAddress[]> favoredNodesSupplier;
+  private final ColumnFamilyDescriptor family;
+  private final Path familyStoreDirectoryPath;
+  private final RegionCoprocessorHost coprocessorHost;
+
+  private StoreContext(Builder builder) {
+    this.blockSize = builder.blockSize;
+    this.encryptionContext = builder.encryptionContext;
+    this.cacheConf = builder.cacheConf;
+    this.regionFileSystem = builder.regionFileSystem;
+    this.comparator = builder.comparator;
+    this.bloomFilterType = builder.bloomFilterType;
+    this.compactedFilesSupplier = builder.compactedFilesSupplier;
+    this.favoredNodesSupplier = builder.favoredNodesSupplier;
+    this.family = builder.family;
+    this.familyStoreDirectoryPath = builder.familyStoreDirectoryPath;
+    this.coprocessorHost = builder.coprocessorHost;
+  }
+
+  public int getBlockSize() {
+    return blockSize;
+  }
+
+  public Encryption.Context getEncryptionContext() {
+    return encryptionContext;
+  }
+
+  public CacheConfig getCacheConf() {
+    return cacheConf;
+  }
+
+  public HRegionFileSystem getRegionFileSystem() {
+    return regionFileSystem;
+  }
+
+  public CellComparator getComparator() {
+    return comparator;
+  }
+
+  public BloomType getBloomFilterType() {
+    return bloomFilterType;
+  }
+
+  public Supplier<Collection<HStoreFile>> getCompactedFilesSupplier() {
+    return compactedFilesSupplier;
+  }
+
+  public InetSocketAddress[] getFavoredNodes() {
+    return favoredNodesSupplier.get();
+  }
+
+  public ColumnFamilyDescriptor getFamily() {
+    return family;
+  }
+
+  public Path getFamilyStoreDirectoryPath() {
+    return familyStoreDirectoryPath;
+  }
+
+  public RegionCoprocessorHost getCoprocessorHost() {
+    return coprocessorHost;
+  }
+
+  public static Builder getBuilder() {
+    return new Builder();
+  }
+
+  @Override
+  public long heapSize() {
+    return FIXED_OVERHEAD;
+  }
```

Comment:
@apurtell ping again on the `heapSize()` and implements `HeapSize` for StoreContext class, could you have another look? I should have fixed it, but will wait for your approval/resolve in 1 day or 2 day before merging it.

Reply:
@apurtell sorry for pinging again, I will merge this evening if I don't hear from you